README.md
# Open Source Arras

**Open Source Arras is beta software.** This build is **not** representative of the final product. Expect bugs and missing features.

## Setup Guide (Localhost)

This guide covers setting up your server on your own hardware and only supports PCs running up-to-date versions of Windows/macOS/Linux.

You'll first need to install [Node.js](https://nodejs.org). It doesn't matter if you pick the LTS or Latest version, they'll both work fine.

Once `Node.js` is installed, run the command `npm i ws`. This will install the WebSocket library that Open Source Arras uses.

After installing `ws`, [download the source code of the latest release of Open Source Arras](https://github.com/Taureon/aps-plus-plus/releases). Extract it once it's downloaded and open either `run.bat` (if you're on Windows) or `run.sh` (if you're not). If there aren't any errors, your server will start up. Go to `localhost:26301` in your favourite web browser (keep the terminal window open, closing it will shut down the server) to play.

[If you need a more detailed guide, click here for a step by step list.](https://github.com/Taureon/aps-plus-plus/wiki/Frequently-Asked-Questions#how-do-i-set-up-my-server)

If you want to stay up to date, fork this template, download a git client, and sync the fork whenever there's a major update.

## Setup Guide (Webhost)

Don't have a supported device or don't want to mess around with localhost? Get a webhost to do the dirty work for you.

Create a new project and choose to import one from GitHub. When prompted for the URL of the repository, type in `https://github.com/Taureon/aps-plus-plus.git`.

Navigate to `server/config.js` and replace `localhost:26301` with the URL for your project. (For Glitch users, it's `your-project.glitch.me` (replace `your-project` with the actual name of your project, it should be above settings)).

For Glitch specifically, go to `package.json` and replace `"node": "18.x"` with `"node": "16.x"`. Glitch doesn't currently support Node 18.

After doing that, your server should be ready!

## Useful Tools
- [Create a custom shape](https://arras.io/ext/custom-shape)
- [Create a custom tank](https://zyrafaq.com/arras-tank-builder) By [DogeisCut](https://github.com/DogeisCut) and [Zyrafaq](https://github.com/zyrafaq)
- [Create a custom theme](https://codepen.io/road-to-100k/full/GRpvMzb)
- [Random Tank Generator](https://perchance.org/chomp-arras-gen)
- [Addon list](https://github.com/Taureon/aps-plus-plus-addons)

## Other Links
- [Our Discord server](https://discord.gg/kvCAZfUCjy)

*p.s. if something goes terribly wrong its not our fault*

changes
exports.laser = {
  PARENT: ["bullet"],
  SHAPE: -1,
  BODY: {
        PENETRATION: 0.7,
        SPEED: 5.2,
        RANGE: 100,
        DENSITY: 0.85,
        HEALTH: 0.1,
        DAMAGE: 4.65,
    }
}
exports.revolutionist = {
    PARENT: ["genericTank"],
    LABEL: "Revolutionist",
    DANGER: 6,
    GUNS: [{
        POSITION: [20, 8, 1, 0, 0, 0, 0],
        PROPERTIES: {
            SHOOT_SETTINGS: combineStats([g.basic, g.flank]),
            TYPE: "bullet",
        },
    },
          ],
    TURRETS: [{
        POSITION: [34, 0, 0, 0, 360, 0],
        TYPE: "turretBase",
    },
  ],
};
exports.revoception = makeCeption(exports.revolutionist, "revoception");
exports.hivemind = {
  PARENT: ["genericTank"],
  LABEL: "Hivemind",
  //CONTROLLERS: ['nearestDifferentMaster'],
  GUNS: [
    {
      /*** LENGTH  WIDTH   ASPECT    X       Y     ANGLE   DELAY */
      POSITION: [18, 8, 1, 0, 0, 0, 0],
      PROPERTIES: {
        SHOOT_SETTINGS: combineStats([g.basic]),
        TYPE: "bullet"
      }
    },
    {
      POSITION: [0, 20, 1, 0, 0, 90, 3],
      PROPERTIES: {
        SHOOT_SETTINGS: combineStats([g.basic, g.slow]),
        TYPE: "hiveprobe",
        MAX_CHILDREN: 1
      }
    },
    {
      POSITION: [0, 20, 1, 0, 0, 270, 3.5],
      PROPERTIES: {
        SHOOT_SETTINGS: combineStats([g.basic, g.slow]),
        TYPE: "hiveprobe",
        MAX_CHILDREN: 1
      }
    }
  ],
  TURRETS: [
    {
      /*  SIZE     X       Y     ANGLE    ARC */
      POSITION: [25, 0, 0, 0, 360, 0],
      TYPE: "mindindicator"
    }
  ]
};
exports.cloner = {
  PARENT: ["genericTank"],
  LABEL: "Cloner",
  //CONTROLLERS: ['nearestDifferentMaster'],
  GUNS: [
    {
      /*** LENGTH  WIDTH   ASPECT    X       Y     ANGLE   DELAY */
      POSITION: [18, 8, 1, 0, 0, 0, 0],
      PROPERTIES: {
        SHOOT_SETTINGS: combineStats([g.basic]),
        TYPE: "bullet"
      }
    },
    {
      POSITION: [0, 20, 1, 0, 0, 180, 3],
      PROPERTIES: {
        SHOOT_SETTINGS: combineStats([g.basic, g.slow]),
        TYPE: "clonerprobe",
        MAX_CHILDREN: 1
      }
    }
  ],
  TURRETS: [
    {
      /*  SIZE     X       Y     ANGLE    ARC */
      POSITION: [22, 0, 0, 0, 360, 0],
      TYPE: "mindindicator"
    }
  ]
};



exports.hiveprobe = {
  PARENT: ["genericTank"],
  LABEL: "Probe",
  TYPE: "drone",
  HITS_OWN_TYPE: "hardWithBuffer",
  DRAW_HEALTH: true,
  BODY: {
    FOV: 0.5,
    SPEED: 3,
    ACCELERATION: 0.4,
    HEALTH: 2.5,
    SHIELD: 0,
    DAMAGE: 1.2,
    RESIST: 1,
    PENETRATION: 1,
    DENSITY: 0.4
  },
  AI: {
    BLIND: true
  },
  CLEAR_ON_MASTER_UPGRADE: true,
  GIVE_KILL_MESSAGE: false,
  ACCEPTS_SCORE: false,
  CONTROLLERS: [
    "nearestDifferentMaster",
    "mapAltToFire",
    "minion",
    "canRepel",
    "hangOutNearMaster"
  ],
  GUNS: [
    {
      /*** LENGTH  WIDTH   ASPECT    X       Y     ANGLE   DELAY */
      POSITION: [18, 8, 1, 0, 0, 0, 0],
      PROPERTIES: {
        SHOOT_SETTINGS: combineStats([g.basic, g.lowpower]),
        TYPE: "bullet"
      }
    }
  ],
  TURRETS: [
    {
      /*  SIZE     X       Y     ANGLE    ARC */
      POSITION: [25, 0, 0, 0, 360, 0],
      TYPE: "mindindicator"
    }
  ]
};
exports.clonerprobe = {
  PARENT: ["hiveprobe"],
  GUNS: [
    {
      /*** LENGTH  WIDTH   ASPECT    X       Y     ANGLE   DELAY */
      POSITION: [18, 8, 1, 0, 0, 0, 0],
      PROPERTIES: {
        SHOOT_SETTINGS: combineStats([g.basic, g.lowpower]),
        TYPE: "bullet"
      }
    }
  ],
    TURRETS: [
    {
      /*  SIZE     X       Y     ANGLE    ARC */
      POSITION: [22, 0, 0, 0, 360, 0],
      TYPE: "mindindicator"
    }
  ]
}

//wait whats stuff
exports.placeableWall = {
    PARENT: ["rock"],
    LABEL: "Wall",
    SIZE: 30,
    SHAPE: 4,
    CLEAR_ON_MASTER_UPGRADE: true,
  TEAM: TEAM_ENEMIES,
      VARIES_IN_SIZE: false,
};
exports.placeableWallSmall = {
    PARENT: ["rock"],
    LABEL: "Wall",
    SIZE: 15,
    SHAPE: 4,
    CLEAR_ON_MASTER_UPGRADE: true,
  TEAM: TEAM_ENEMIES,
      VARIES_IN_SIZE: false,
};
exports.wallPlacerThing = {
    PARENT: ["genericTank"],
    SHAPE:0,
    MIRROR_MASTER_ANGLE: true,
    INTANGIBLE: true,
  DRAW_SELF:false,
    COLOR: 16,
      CLEAR_ON_MASTER_UPGRADE: true,
      BODY: {
        ACCELERATION: 0.1,
        SPEED: true,
        HEALTH: 340282366920938463463374607431768211455,
        RESIST: 1,
        SHIELD: 340282366920938463463374607431768211455,
        REGEN: 340282366920938463463374607431768211455,
        DAMAGE: false,
        PENETRATION: true,
        RANGE: true,
        FOV: true,
        SHOCK_ABSORB: 340282366920938463463374607431768211455,
        RECOIL_MULTIPLIER: false,
        DENSITY: 340282366920938463463374607431768211455,
        STEALTH: true,
        PUSHABILITY: false,
        HETERO: false,
    },
    MOTION_TYPE: "withMouse",
   GUNS: [
       {
            POSITION: [0, 20, 1, 10, 0, 0, 0],
            PROPERTIES: {
                SHOOT_SETTINGS: combineStats([{reload:10, speed:0, maxSpeed:0, shudder:0.0001, spray:0.0001}]),
                TYPE: "placeableWall",
                COLOR: 16,
                LABEL: "",
                STAT_CALCULATOR: 0,
                WAIT_TO_CYCLE: false,
                AUTOFIRE: false,
                SYNCS_SKILLS: false,
                MAX_CHILDREN: 0,
                ALT_FIRE: false,
                NEGATIVE_RECOIL: false,
              DRAW_FILL:false,
              BORDERLESS:true,
            },
        },
    ],

};
exports.wallPlacer = {
    PARENT: ["genericTank"],
    LABEL: "Messin' Around",
    BODY: {
        ACCELERATION: base.ACCEL * 1,
        SPEED: base.SPEED * 1,
        HEALTH: base.HEALTH * 1,
        DAMAGE: base.DAMAGE * 1,
        PENETRATION: base.PENETRATION * 1,
        SHIELD: base.SHIELD * 1,
        REGEN: base.REGEN * 1,
        FOV: base.FOV * 1,
        DENSITY: base.DENSITY * 1,
        PUSHABILITY: 1,
        HETERO: 3,
    },
    GUNS: [
       {
            POSITION: [16, 20, 1, 0, 0, 0, 0],

server/index.js
const path = require('path');
const fs = require('fs');

Error.stackTraceLimit = Infinity;
let enviroment = require('./lib/dotenv.js')(fs.readFileSync(path.join(__dirname, '../.env')).toString());
for (let key in enviroment) {
    process.env[key] = enviroment[key];
}
const GLOBAL = require("./modules/global.js");

console.log(`[${GLOBAL.creationDate}]: Server initialized.\nRoom Info:\n Dimensions: ${room.width} x ${room.height}`);

// Let's get a cheaper array removal thing
Array.prototype.remove = function (index) {
    if (index === this.length - 1) return this.pop();
    let r = this[index];
    this[index] = this.pop();
    return r;
};

//console window title
// https://stackoverflow.com/questions/29548477/how-do-you-set-the-terminal-tab-title-from-node-js
process.stdout.write(String.fromCharCode(27) + "]0;" + c.WINDOW_NAME + String.fromCharCode(7));

util.log(room.width + " x " + room.height + " room initalized.");

// Collision stuff
function collide(collision) {
    // Pull the two objects from the collision grid
    let instance = collision[0],
        other = collision[1];
    if (instance.onDef != null) instance.ON(undefined, 'collide', { instance, other })
    // Check for ghosts...
    if (other.isGhost) {
        util.error("GHOST FOUND");
        util.error(other.label);
        util.error("x: " + other.x + " y: " + other.y);
        util.error(other.collisionArray);
        util.error("health: " + other.health.amount);
        util.warn("Ghost removed.");
        if (grid.checkIfInHSHG(other)) {
            util.warn("Ghost removed.");
            grid.removeObject(other);
        }
        return 0;
    }
    if (instance.isGhost) {
        util.error("GHOST FOUND");
        util.error(instance.label);
        util.error("x: " + instance.x + " y: " + instance.y);
        util.error(instance.collisionArray);
        util.error("health: " + instance.health.amount);
        if (grid.checkIfInHSHG(instance)) {
            util.warn("Ghost removed.");
            grid.removeObject(instance);
        }
        return 0;
    }
    if (
        (!instance.activation.check() && !other.activation.check()) ||
        (instance.ac && !instance.alpha) ||
        (other.ac && !other.alpha)
    ) return 0;
    switch (true) {
        case instance.type === "wall" || other.type === "wall":
            if (instance.type === "wall" && other.type === "wall") return;
            if (instance.type === "aura" || other.type === "aura") return;
            let wall = instance.type === "wall" ? instance : other;
            let entity = instance.type === "wall" ? other : instance;
            if (entity.ac || entity.master.ac) return;
            switch (wall.shape) {
                case 4:
                    reflectCollide(wall, entity);
                    break;
                case 0:
                    mooncollide(wall, entity);
                    break;
                default:
                    let a = entity.type === "bullet" ? 1 + 10 / (entity.velocity.length + 10) : 1;
                    advancedcollide(wall, entity, false, false, a);
                    break;
            }
            break;
        case instance.team === other.team &&
            (instance.settings.hitsOwnType === "pushOnlyTeam" ||
                other.settings.hitsOwnType === "pushOnlyTeam"):
            {
                let pusher = instance.settings.hitsOwnType === "pushOnlyTeam" ? instance : other;
                let entity = instance.settings.hitsOwnType === "pushOnlyTeam" ? other : instance;
                // Dominator / Mothership collisions
                if (
                    instance.settings.hitsOwnType === other.settings.hitsOwnType ||
                    entity.settings.hitsOwnType === "never"
                ) return;
                let a = 1 + 10 / (Math.max(entity.velocity.length, pusher.velocity.length) + 10);
                advancedcollide(pusher, entity, false, false, a);
            }
            break;
        case (instance.type === "crasher" && other.type === "food") ||
            (other.type === "crasher" && instance.type === "food"):
            firmcollide(instance, other);
            break;
        case instance.team !== other.team ||
            (instance.team === other.team &&
            (
                instance.healer ||
                other.healer
            )):
            // Exits if the aura is not hitting a boss or tank
            if (instance.type === "aura") {
                if (!(other.type === "tank" || other.type === "miniboss" || other.type == "food")) return;
            } else if (other.type === "aura") {
                if (!(instance.type === "tank" || instance.type === "miniboss" || instance.type == "food")) return;
            }
            advancedcollide(instance, other, true, true);
            break;
        case instance.settings.hitsOwnType == "never" ||
            other.settings.hitsOwnType == "never":
            break;
        case instance.settings.hitsOwnType === other.settings.hitsOwnType:
            switch (instance.settings.hitsOwnType) {
                case 'assembler': {
                    if (instance.assemblerLevel == null) instance.assemblerLevel = 1;
                    if (other.assemblerLevel == null) other.assemblerLevel = 1;

                    const [target1, target2] = (instance.id > other.id) ? [instance, other] : [other, instance];

                    if (
                        target2.assemblerLevel >= 10 || target1.assemblerLevel >= 10 ||
                        target1.isDead() || target2.isDead() ||
                        target1.parent.id != target2.parent.id &&
                        target1.parent.id != null &&
                        target2.parent.id != null // idk why
                    ) {
                        advancedcollide(instance, other, false, false); // continue push
                        break;
                    }

                    const better = (state) => {
                        return target1[state] > target2[state] ? target1[state] : target2[state];
                    }

                    target1.assemblerLevel = Math.min(target2.assemblerLevel + target1.assemblerLevel, 10);
                    target1.SIZE = better('SIZE') * 1.1;
                    target1.SPEED = better('SPEED') * 0.9;
                    target1.HEALTH = better('HEALTH') * 1.2;
                    target1.health.amount = target1.health.max;
                    target1.DAMAGE = better('DAMAGE') * 1.1;
                    target2.kill();

                    for (let i = 0; i < 10; ++i) {
                        const o = new Entity(target1, target1);
                        o.define('assemblerEffect');
                        o.team = target1.team;
                        o.color = target1.color;
                        o.SIZE = target1.SIZE / 3;
                        o.velocity = new Vector((Math.random() - 0.5) * 25, (Math.random() - 0.5) * 25);
                        o.refreshBodyAttributes();
                        o.life();
                    }
                } // don't break
                case "push":
                    advancedcollide(instance, other, false, false);
                    break;
                case "hard":
                    firmcollide(instance, other);
                    break;
                case "hardWithBuffer":
                    firmcollide(instance, other, 30);
                    break;
                case "hardOnlyTanks":
                    if (
                        instance.type === "tank" &&
                        other.type === "tank" &&
                        !instance.isDominator &&
                        !other.isDominator
                    )
                        firmcollide(instance, other);
                case "hardOnlyBosses":
                    if (instance.type === other.type && instance.type === "miniboss")
                        firmcollide(instance, other);
                case "repel":
                    simplecollide(instance, other);
                    break;
            }
            break;
    }
}

// The most important loop. Lots of looping.
let time, ticks = 0;
const gameloop = () => {
    logs.loops.tally();
    logs.master.set();
    logs.activation.set();
    logs.activation.mark();
    // Do collisions
    logs.collide.set();
    if (entities.length > 1) {
        // Load the grid
        grid.update();
        // Run collisions in each grid
        const pairs = grid.queryForCollisionPairs();
        for (let i = 0; i < pairs.length; i++) {
            collide(pairs[i]);
        }
    }
    logs.collide.mark();
    // Do entities life
    logs.entities.set();
    for (let my of entities) {
        // Consider death.
        if (my.contemplationOfMortality()) {
            my.destroy();
        } else {
            if (my.bond == null) {
                // Resolve the physical behavior from the last collision cycle.
                logs.physics.set();
                my.physics();
                logs.physics.mark();
            }
            if (my.activation.check() || my.isPlayer) {
                logs.entities.tally();
                // Think about my actions.
                logs.life.set();
                my.life();
                logs.life.mark();
                // Apply friction.
                poison(my);
                freeze(my);
                my.friction();
                my.confinementToTheseEarthlyShackles();
                logs.selfie.set();
                my.takeSelfie();
                logs.selfie.mark();
            }
            // Update collisions.
            my.collisionArray = [];
            // Activation
            my.activation.update();
            my.updateAABB(my.activation.check());
        }
        // Update collisions.
        my.collisionArray = [];
        if (my.onDef != null) my.ON(undefined, 'tick')
    }
    logs.entities.mark();
    logs.master.mark();
    // Remove dead entities
    purgeEntities();
    room.lastCycle = util.time();
    ticks++;
    if (ticks & 1) {
        for (let i = 0; i < sockets.players.length; i++) {
            sockets.players[i].socket.view.gazeUpon();
            sockets.players[i].socket.lastUptime = Infinity;
        }
    }
};

setTimeout(closeArena, 60000 * 120); // Restart every 2 hours

global.naturallySpawnedBosses = [];
global.bots = [];
let bossTimer = 0;
// A less important loop.
let maintainloop = () => {
    // Regen health and update the grid
    for (let i = 0; i < entities.length; i++) {
        let instance = entities[i];
        if (instance.shield.max) {
            instance.shield.regenerate();
        }
        if (instance.health.amount) {
            instance.health.regenerate(instance.shield.max && instance.shield.max === instance.shield.amount);
        }
    }
    if (!naturallySpawnedBosses.length && bossTimer++ > c.BOSS_SPAWN_COOLDOWN) {
        bossTimer = -c.BOSS_SPAWN_DURATION;
        let selection = c.BOSS_TYPES[ran.chooseChance(...c.BOSS_TYPES.map((selection) => selection.chance))],
            amount = ran.chooseChance(...selection.amount) + 1;
        if (selection.message) {
            sockets.broadcast(selection.message);
        }
        sockets.broadcast(amount > 1 ? "Visitors are coming." : "A visitor is coming.");
        setSyncedTimeout(() => {
            let names = [];

            for (let i = 0; i < amount; i++) {
                let spot, attempts = 30, name = ran.chooseBossName(selection.nameType);
                do { spot = getSpawnableArea(TEAM_ENEMIES); } while (attempts-- && dirtyCheck(spot, 500));

                let boss = new Entity(spot);
                boss.define(selection.bosses.sort(() => 0.5 - Math.random())[i % selection.bosses.length]);
                boss.team = TEAM_ENEMIES;
                if (name) {
                    boss.name = name;
                }

                names.push(boss.name);
                naturallySpawnedBosses.push(boss);
                boss.on('dead', () => util.remove(naturallySpawnedBosses, naturallySpawnedBosses.indexOf(boss)));
            }

            sockets.broadcast(`${util.listify(names)} ${names.length == 1 ? 'has' : 'have'} arrived!`);
        }, c.BOSS_SPAWN_DURATION * 30);
    }

    // upgrade existing ones
    for (let i = 0; i < bots.length; i++) {
        let o = bots[i];
        if (o.skill.level < c.LEVEL_CAP) {
            o.skill.score += c.BOT_XP;
        }
        o.skill.maintain();
        o.skillUp([ "atk", "hlt", "spd", "str", "pen", "dam", "rld", "mob", "rgn", "shi" ][ran.chooseChance(...c.BOT_SKILL_UPGRADE_CHANCES)]);
        if (o.leftoverUpgrades && o.upgrade(ran.irandomRange(0, o.upgrades.length))) {
            o.leftoverUpgrades--;
        }
    }

    // then add new bots if arena is open
    if (!global.arenaClosed && bots.length < c.BOTS) {
        let team = c.MODE === "tdm" ? getWeakestTeam() : undefined,
            limit = 20, // give up after 20 attempts and just pick whatever is currently chosen
            loc;
        do {
            loc = getSpawnableArea(team);
        } while (limit-- && dirtyCheck(loc, 50))
        let o = new Entity(loc);
        o.define('bot');
        o.define(c.SPAWN_CLASS);
        o.refreshBodyAttributes();
        o.isBot = true;
        o.name += ran.chooseBotName();
        o.leftoverUpgrades = ran.chooseChance(...c.BOT_CLASS_UPGRADE_CHANCES);
        o.color = c.RANDOM_COLORS ? Math.floor(Math.random() * 20) : team ? getTeamColor(team) : 17;
        if (team) o.team = team;
        bots.push(o);
        o.on('dead', () => util.remove(bots, bots.indexOf(o)));
    }
};

//evaluating js with a seperate console window if enabled
if (c.REPL_WINDOW) {
    util.log('Starting REPL Terminal.');
    //TODO: figure out how to spawn a seperate window and put the REPL stdio in there instead
    //let { stdin, stdout, stderr } = (require('child_process').spawn("cmd.exe", ["/c", "node", "blank.js"], { detached: true }));
    require('repl').start({/* stdin, stdout, stderr,*/ useGlobal: true });
}

// Bring it to life
let counter = 0;
setInterval(() => {
    gameloop()
    gamemodeLoop();
    roomLoop();

    if (counter++ / c.runSpeed > 30) {
        chatLoop();
        maintainloop();
        speedcheckloop();
        counter = 0;
    }

    syncedDelaysLoop();
}, room.cycleSpeed);
